md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2037 @@
+
+> Validation types are only used during validation, and not during execution.
+
+> `any` is effectively serving as a kind of bottom type here, and `void` is
+effectively serving as a kind of top type here. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Validation Type Merge
+
+To merge two validation types:
+ - If either validation type is `any`, the result is the other validation type.
+ - Otherwise, if the validation types are the same, the result is that
+   validation type.
+ - Otherwise, validation fails.
+
+### Module Validation
+
+TODO: Describe module validation.
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries each containing
+    - A [validation type](#validation-type).
+    - A *limit* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During function validation, if either stack is empty when an element is to be
+popped from it, or an element is accessed by index less than zero or not less
+than the stack's length, validation fails. If a [validation type] is popped from
+the type stack such that the length of the type stack becomes less than the
+control-flow stack top's limit value, validation fails.
+
+#### Function Validation Initialization
+
+If the instruction sequence is empty, or if the last instruction in the sequence
+is not an `end`, validation fails.
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's [validation type] is
+the first return type of the function, if it has any, or `void` otherwise. Its
+limit value is zero.
+
+> The final `end` instruction may be implicit in some representations.
+
+#### Function Body Validation
+
+The instruction at the current position is remembered, and the current position
+is incremented to point to the position following it. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, [generic validation](#generic-instruction-validation) is
+performed.
+
+If the current position is now past the end of the sequence,
+[function return validation](#function-return-validation) is initiated and
+validation of the function is thereafter complete.
+
+Otherwise, [validation](#function-body-validation) is restarted with the new
+current position.
+
+#### Generic Instruction Validation
+
+For each operand [type] in the instruction's signature in reverse order, a
+[validation type] is popped from the type stack and checked to match it. Each
+return type of the instruction's signature is then pushed onto the type stack.
+
+If the instruction's signture has no return types:
+ - If the length of the type stack is greater than the control-flow stack top's
+   limit value, validation fails.
```

Comment:
Ok, I think I understand the proposal now, thank you. But this seems a significant change, and would block multiple use of values on the stack, block calls pushing multiple values on the stack, a very limiting constraint for wasm. It's fine for an expressionless encoding as then there are no values on a values stack, and this seems to be where this is heading, to a heavy dependence on local variables which might be fine.

Adding back the pushing of the `void` type will also break the potential multiple value call support. It will require multiple value results be written directly to local variables rather than pushed on the stack, for example:

```
call $fn_returning_two_values : pushing two values on the stack.
set_local $l1 : popping one value, and pushing void.
set_local $l2 : fails because either popping the void or the stack is not empty.
```
This would fail to validate under either constraint. Either the stack is not empty or a `void` type is push and the first call result push could no be consumed.

I don't think these are even compatible constraints: as then `(block (nop) (nop))` would be invalid as the first pushes `void`.